### PR TITLE
(WIP): Add Warmup and Cooldown Periods

### PR DESCRIPTION
**Pull request by coachchucksol**
Adding in warmup and cooldown periods to address: Issue #17

### CODE REVIEW

File: unit_tests/tests/core/slot_toggled_field.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#[cfg(test)]
```
Comment (buffalu) [inline unit_tests/tests/core/slot_toggled_field.rs:1]: you can throw all these into the slot_toggled_field file at the bottom

File: core/src/slot_toggled_field.rs [hunk 2]
```diff
@@ -24,25 +32,62 @@ impl SlotToggle {
...
+    }
+
+    pub fn cooldown_slots(&self, slot: u64, epoch_length: u64) -> u64 {
+        let remaining_slots = slot.checked_rem(epoch_length).unwrap_or(0);
+        self.slot_removed
+            .saturating_add(epoch_length)
+            .saturating_add(remaining_slots)
+    }
+
+    #[allow(clippy::collapsible_else_if)]
```
Comment (buffalu) [inline core/src/slot_toggled_field.rs:77]: should fix that

File: core/src/slot_toggled_field.rs [hunk 3]
```diff
@@ -24,25 +32,62 @@ impl SlotToggle {
...
+    pub fn state(&self, slot: u64, epoch_length: u64) -> SlotToggleState {
+        if self.slot_added >= self.slot_removed {
+            if slot <= self.warmup_slots(slot, epoch_length) {
```
Comment (buffalu) [inline core/src/slot_toggled_field.rs:80]:
```
let slot_added_epoch = slot.checked_div(epoch_length).unwrap();
let current_epoch = slot.checked_div(epoch_length).unwrap();
if current_epoch > slot_added_epoch + 1 {
    SlotToggleState::Active
} else {
    SlotToggleState::WarmUp
}
```

File: core/src/slot_toggled_field.rs [hunk 4]
```diff
@@ -24,25 +32,62 @@ impl SlotToggle {
...
+                self.slot_removed = slot;
+                true
+            }
+            _ => false,
         }
     }
 
-    pub const fn is_active(&self, slot: u64) -> bool {
-        self.slot_added >= self.slot_removed && slot >= self.slot_added
+    pub fn is_active(&self, slot: u64, epoch_length: u64) -> bool {
```
Comment (buffalu) [inline core/src/slot_toggled_field.rs:55]: perhaps is_staked_or_slashable? is_active conflicts with state. open to other ideas

File: unit_tests/tests/core/slot_toggled_field.rs (hunk 1 above)
Comment (coachchucksol) [reply unit_tests/tests/core/slot_toggled_field.rs:1]: Can do

File: core/src/slot_toggled_field.rs (hunk 3 above)
Comment (coachchucksol) [reply core/src/slot_toggled_field.rs:80]: Done

File: core/src/slot_toggled_field.rs (hunk 2 above)
Comment (coachchucksol) [reply core/src/slot_toggled_field.rs:77]: Done

File: core/src/slot_toggled_field.rs [hunk 5]
```diff
@@ -24,25 +32,109 @@ impl SlotToggle {
...
-        if self.slot_added < self.slot_removed {
-            false
+    pub fn is_staked_or_slashable(&self, slot: u64, epoch_length: u64) -> bool {
+        matches!(
+            self.state(slot, epoch_length),
+            SlotToggleState::Active | SlotToggleState::Cooldown
+        )
+    }
+
+    pub fn is_active(&self, slot: u64, epoch_length: u64) -> bool {
```
Comment (buffalu) [inline core/src/slot_toggled_field.rs:62]: imo just have this return matches! State::Active

File: vault_program/src/slash.rs [hunk 6]
```diff
@@ -54,31 +55,35 @@ pub fn process_slash(
...
     vault_operator_ticket
         .vault_operator_ticket()
-        .check_active(slot)?;
+        .check_active(slot, config.config().epoch_length())?;
 
     // The operator shall be opted-in to the AVS and the AVS shall be opted-in to the operator
     avs_operator_ticket
         .avs_operator_ticket()
-        .check_active(slot)?;
+        .check_active(slot, config.config().epoch_length())?;
```
Comment (buffalu) [inline vault_program/src/slash.rs:76]: check_staked for these, which would be active or cooling down

File: core/src/slot_toggled_field.rs (hunk 5 above)
Comment (coachchucksol) [reply core/src/slot_toggled_field.rs]: Implemented what we talked about

File: vault_program/src/slash.rs (hunk 6 above)
Comment (coachchucksol) [reply vault_program/src/slash.rs]: Implemented what we talked about

File: core/src/slot_toggled_field.rs [hunk 7]
```diff
@@ -24,25 +32,101 @@ impl SlotToggle {
...
+            self.state(slot, epoch_length),
+            SlotToggleState::Active | SlotToggleState::Cooldown
+        )
+    }
+
+    pub fn state(&self, slot: u64, epoch_length: u64) -> SlotToggleState {
+        let current_epoch = slot.checked_div(epoch_length).unwrap();
+
+        if self.slot_added >= self.slot_removed {
+            let slot_added_epoch = self.slot_added().checked_div(epoch_length).unwrap();
```
Comment (buffalu) [inline core/src/slot_toggled_field.rs:66]: ultra nit: self.slot_added above + function here. same below

**Review by buffalu — APPROVED** [review]

**Pull request merged**